ccl/sqlproxyccl: idle connection timeout support

darinpp opened the pull request:
Previusly, the connection from the end user to the backend
can be idle for any period of time without disconnecting.
In certain cases, we want the ability for idle connections
to disconnect when the idle time exceeds redefined timeout.
To address this, this patch adds tracking of how long
the connection to the backend has been idle. If there is
a timeout specified, the connection will be disconnected
and an error message will be sent back to the end user.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/57431)
<!-- Reviewable:end -->

spaskob reviewed (commented):
Thanks - looks like quite an elegant approach!
Maybe I am missing some context but why idle terminate the backend connection vs the client connection? In practice probably it does not matter but I would have thought that we are trying to prevent a client opening a connection and letting stay open for a long time without any activity on it. Also, I don't understand at all the pg wire protocol but are there any liveness pings that would keep the connection open? Is there a way to set idle timeout on pgx directly?

darinpp commented on pkg/ccl/sqlproxyccl/proxy.go:
```diff
@@ -32,6 +33,9 @@ type BackendConfig struct {
 	OnConnectionSuccess func()
+	// Idle connection timeout. If set - will cause the idle backend connections
+	// to timeout after waiting to read or write for that duration.
+	IdleTimeout time.Duration
```
In the ideal future state, we won't terminate the connection to the client. When we detect an idle connection, the proxy will initiate serialization of the tenant's state, disconnect the connection to the tenant and shutdown the tenant pod. When a new command comes from the client, the proxy will transparently establish a new backend connection to a new pod, stamp it with the tenant id, deserialize the state and continue as if the tenant pod never terminated.

spaskob replied:
thanks, make senses now

darinpp commented on pkg/ccl/sqlproxyccl/proxy.go:
```diff
@@ -55,6 +59,9 @@ type Options struct {
 	// The error passed to this method will contain no internal information.
 	OnSendErrToClient func(code ErrorCode, msg string) string
+
+	// If set, will be used to establish and return connection to the backend.
```
As long as `NewServer` is used - it would be initialized and not nil.

spaskob replied:
I meant we probably need a comment on what Dialer is used when none is provided.

darinpp commented on pkg/ccl/sqlproxyccl/error.go:
```diff
@@ -62,20 +62,25 @@ const (
 
-type codeError struct {
+// CodeError bundle an error with an error code
+type CodeError struct {
```
The linter complains if it isn't. `BackendDialer` is returning it so the caller can check the code (vs the error text) and act on it.

spaskob commented on pkg/ccl/sqlproxyccl/idle_disconnect_connection.go:
```diff
@@ -0,0 +1,61 @@
+
+// Write writes data to the connection and sets the read timeout.
+func (c IdleDisconnectConnection) Write(b []byte) (n int, err error) {
+	if err := c.Conn.SetReadDeadline(timeutil.Now().Add(c.timeout)); err != nil {
```
Using a read deadline is not obvious, add a comment perhaps?

darinpp replied:
added.

darinpp commented on pkg/ccl/sqlproxyccl/proxy.go:
```diff
@@ -152,39 +166,19 @@ func (s *Server) Proxy(proxyConn *Conn) error {
-		if response[0] != pgAcceptSSLRequest {
-			s.metrics.BackendDownCount.Inc(1)
-			return NewErrorf(CodeBackendRefusedTLS, "target server refused TLS connection")
+	crdbConn, errWithCode := s.backendDialer(backendConfig)
```
True. I removed the field from `Server`

darinpp commented on pkg/ccl/sqlproxyccl/proxy.go:
```diff
@@ -219,6 +213,11 @@ func (s *Server) Proxy(proxyConn *Conn) error {
 	// client gets to close the connection once it's sent that message,
 	// meaning either case is possible.
 	case err := <-errIncoming:
```
Moved the `err == nil` check first

darinpp commented on pkg/ccl/sqlproxyccl/proxy.go:
```diff
@@ -152,39 +166,19 @@ func (s *Server) Proxy(proxyConn *Conn) error {
+	if s.opts.ModifyRequestParams != nil {
+		s.opts.ModifyRequestParams(msg.Parameters)
 	}
 	defer func() { _ = crdbConn.Close() }()
```
Yeah. The merge seems to have moved it. Fixed.

darinpp commented on pkg/ccl/sqlproxyccl/error.go:
```diff
@@ -62,20 +62,25 @@ const (
 )
 
-type codeError struct {
+// CodeError bundle an error with an error code
```
Done

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57431)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @darinpp, @imjching, and @spaskob)

___
*[pkg/ccl/sqlproxyccl/backend_dialer.go, line 25 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MNfDflA4EsEfRal1Bv3:-MNfDflA4EsEfRal1Bv4:blipdz4) ([raw file](https://github.com/cockroachdb/cockroach/blob/f36136b1cdb61053b5caff6686e1c53aff0147a0/pkg/ccl/sqlproxyccl/backend_dialer.go#L25)):*
> ```Go
> 	}
> 
> 	if backendConfig.TLSConf != nil {
> ```

This doesn't seem quite right to me design-wise. The dialer should just establish the connection, and the proxy should then send the SSL requests and wrap with the idle timeout connection. Otherwise, if I provide my own BackendDialer, I'd have to do those actions myself. In other words, the dialer just establishes the raw TCP connection, and it's up to the proxy to do the PG stuff on top of that.

This also allows us to simplify the dialer function to be:
```
func(backendConfig *BackendConfig) (net.Conn, error)
```
Since if any error occurs during connection, it will get mapped to `CodeBackendDown` by `sqlproxyccl`.
___
*[pkg/ccl/sqlproxyccl/idle_disconnect_connection.go, line 28 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MNfBznZ7ueKScr7qcrn:-MNfBznZ7ueKScr7qcro:bnx5hu1) ([raw file](https://github.com/cockroachdb/cockroach/blob/f36136b1cdb61053b5caff6686e1c53aff0147a0/pkg/ccl/sqlproxyccl/idle_disconnect_connection.go#L28)):*
> ```Go
> 
> // Read reads data from the connection with timeout.
> func (c IdleDisconnectConnection) Read(b []byte) (n int, err error) {
> ```

I think all these methods should use pointer receivers (c *IdleDisconnectConnection), since otherwise there might be unnecessary copying when they are called.
___
*[pkg/ccl/sqlproxyccl/idle_disconnect_connection.go, line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MNfCXaZ2B0RtxIwQVOH:-MNfCXaZ2B0RtxIwQVOI:b-qg3b94) ([raw file](https://github.com/cockroachdb/cockroach/blob/f36136b1cdb61053b5caff6686e1c53aff0147a0/pkg/ccl/sqlproxyccl/idle_disconnect_connection.go#L29)):*
> ```Go
> // Read reads data from the connection with timeout.
> func (c IdleDisconnectConnection) Read(b []byte) (n int, err error) {
> 	if err := c.Conn.SetReadDeadline(timeutil.Now().Add(c.timeout)); err != nil {
> ```

Hm, I feel uncertain about the perf impact of calling `Now` and `SetReadDeadline` on every single `Read` and `Write` call. That makes me nervous, rather than using a lower overhead mechanism. Can you do a micro-benchmark to see how much overhead this adds?
___
*[pkg/ccl/sqlproxyccl/idle_disconnect_connection.go, line 37 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MNfCsqe9fpU6DQGyjGv:-MNfCsqe9fpU6DQGyjGw:b1y5msd) ([raw file](https://github.com/cockroachdb/cockroach/blob/f36136b1cdb61053b5caff6686e1c53aff0147a0/pkg/ccl/sqlproxyccl/idle_disconnect_connection.go#L37)):*
> ```Go
> // Write writes data to the connection and sets the read timeout.
> func (c IdleDisconnectConnection) Write(b []byte) (n int, err error) {
> 	// The Write for the connection is not blocking (or can block only temporary in case of flow control).
> ```

NIT: In CRDB codebase, comments shouldn't extend past 80 char limit.
___
*[pkg/ccl/sqlproxyccl/proxy.go, line 32 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MNfEKCr2o3cBeVo8eEv:-MNfEKCr2o3cBeVo8eEw:bv8265q) ([raw file](https://github.com/cockroachdb/cockroach/blob/f36136b1cdb61053b5caff6686e1c53aff0147a0/pkg/ccl/sqlproxyccl/proxy.go#L32)):*
> ```Go
> type BackendConfig struct {
> 	// The address to which the connection is forwarded.
> 	OutgoingAddress string
> ```

Can't we get rid of `OutgoingAddress` now that we have a dialer function?
___
*[pkg/ccl/sqlproxyccl/proxy.go, line 60 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MNfFSKZ-q8ADw9F08V8:-MNfFSK_1FoalxTR9dY-:bqifia9) ([raw file](https://github.com/cockroachdb/cockroach/blob/f36136b1cdb61053b5caff6686e1c53aff0147a0/pkg/ccl/sqlproxyccl/proxy.go#L60)):*
> ```Go
> 	// TODO(tbg): this is unimplemented and exists only to check which clients
> 	// allow use of SNI. Should always return ("", nil).
> 	BackendConfigFromSNI func(serverName string) (config *BackendConfig, clientErr error)
> ```

These existing fields look overly complex. I think we'll want another PR where we simplify the callers to just the dialer function, so they don't need some of this other stuff. Adding the dialer function doesn't simplify anything right now; it actually makes it more complex, since it's additive and not simplifying anything else.


<!-- Sent from Reviewable.io -->

darinpp reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57431)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @darinpp, @imjching, and @spaskob)

___
*[pkg/ccl/sqlproxyccl/backend_dialer.go, line 25 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MNfDflA4EsEfRal1Bv3:-MO7cLZl04wexBZOKHAK:b-650g82) ([raw file](https://github.com/cockroachdb/cockroach/blob/f36136b1cdb61053b5caff6686e1c53aff0147a0/pkg/ccl/sqlproxyccl/backend_dialer.go#L25)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

This doesn't seem quite right to me design-wise. The dialer should just establish the connection, and the proxy should then send the SSL requests and wrap with the idle timeout connection. Otherwise, if I provide my own BackendDialer, I'd have to do those actions myself. In other words, the dialer just establishes the raw TCP connection, and it's up to the proxy to do the PG stuff on top of that.

This also allows us to simplify the dialer function to be:
```
func(backendConfig *BackendConfig) (net.Conn, error)
```
Since if any error occurs during connection, it will get mapped to `CodeBackendDown` by `sqlproxyccl`.
</blockquote></details>

This has been changed and now the dialer has this signature `BackendDialer func(*pgproto3.StartupMessage) (net.Conn, error)`. Not every use of the proxy will require SSL to the backend so it is still the backend dialer that is responsible for the SSL setup. But what is provided is a set of bulding block (overlays) so that makes it easy to build a dialer with the desired characteristics. 
___
*[pkg/ccl/sqlproxyccl/idle_disconnect_connection.go, line 28 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MNfBznZ7ueKScr7qcrn:-MO7cw5W01aHNnQIxIpN:b9k5jwh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f36136b1cdb61053b5caff6686e1c53aff0147a0/pkg/ccl/sqlproxyccl/idle_disconnect_connection.go#L28)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I think all these methods should use pointer receivers (c *IdleDisconnectConnection), since otherwise there might be unnecessary copying when they are called.
</blockquote></details>

changed. 
___
*[pkg/ccl/sqlproxyccl/idle_disconnect_connection.go, line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MNfCXaZ2B0RtxIwQVOH:-MO7dJvP1O9GP0_XpyEp:bh2gghe) ([raw file](https://github.com/cockroachdb/cockroach/blob/f36136b1cdb61053b5caff6686e1c53aff0147a0/pkg/ccl/sqlproxyccl/idle_disconnect_connection.go#L29)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Hm, I feel uncertain about the perf impact of calling `Now` and `SetReadDeadline` on every single `Read` and `Write` call. That makes me nervous, rather than using a lower overhead mechanism. Can you do a micro-benchmark to see how much overhead this adds?
</blockquote></details>

Changed so now the call happens once every 15 seconds instead. The benchmark also shows that there is no difference. 
___
*[pkg/ccl/sqlproxyccl/idle_disconnect_connection.go, line 37 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MNfCsqe9fpU6DQGyjGv:-MO7dnfl3pkUK6c6hJNh:b-n9tkhd) ([raw file](https://github.com/cockroachdb/cockroach/blob/f36136b1cdb61053b5caff6686e1c53aff0147a0/pkg/ccl/sqlproxyccl/idle_disconnect_connection.go#L37)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: In CRDB codebase, comments shouldn't extend past 80 char limit.
</blockquote></details>

changed
___
*[pkg/ccl/sqlproxyccl/proxy.go, line 32 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MNfEKCr2o3cBeVo8eEv:-MO7dtiK7A2-o9zxpCCq:b1w4ick) ([raw file](https://github.com/cockroachdb/cockroach/blob/f36136b1cdb61053b5caff6686e1c53aff0147a0/pkg/ccl/sqlproxyccl/proxy.go#L32)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Can't we get rid of `OutgoingAddress` now that we have a dialer function?
</blockquote></details>

We should be able to get rid of `BackendConfig` entirely once everything is switched to use `BackendDialer`. This will happen once the managed services repo is updated. 
___
*[pkg/ccl/sqlproxyccl/proxy.go, line 60 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MNfFSKZ-q8ADw9F08V8:-MO7e9j8EA-pLzRtEmQX:b-688h88) ([raw file](https://github.com/cockroachdb/cockroach/blob/f36136b1cdb61053b5caff6686e1c53aff0147a0/pkg/ccl/sqlproxyccl/proxy.go#L60)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

These existing fields look overly complex. I think we'll want another PR where we simplify the callers to just the dialer function, so they don't need some of this other stuff. Adding the dialer function doesn't simplify anything right now; it actually makes it more complex, since it's additive and not simplifying anything else.
</blockquote></details>

I removed this one and made it so it just adds the SNI server name as param. 


<!-- Sent from Reviewable.io -->

darinpp committed:
ccl/sqlproxyccl: idle connection timout support

Previusly, the connection from the end user to the backend
can be idle for any period of time without disconnecting.
In certain cases, we want the ability for idle connections
to disconnect when the idle time exceeds redefined timeout.
To address this, this patch adds tracking of how long
the connection to the backend has been idle. If there is
a timeout specified, the connection will be disconnected
and an error message will be sent back to the end user.

This PR also refactors the proxy configuration.
Previsouly - the proxy was cnfigured via BackendConfigFromParams
func passed by the proxy user. This is now deprecated and
instead proxy user has to pass a BackendDialer that does that.

Release note: None

darinpp commented:
This is now ready for another review.
As agreed, the `BackendDialer` is now used in all cases except when `OnConnectionSuccess` and `KeepAliveLoop` are needed. 
For example
```
	opts := Options{
		BackendConfigFromParams: func(params map[string]string, _ *Conn) (*BackendConfig, error) {
			return &BackendConfig{
				OnConnectionSuccess: func() { connSuccess = true },
			}, nil
		}, 
		BackendDialer: func(msg *pgproto3.StartupMessage) (net.Conn, error) {
			return BackendDial(
				msg, tc.Server(0).ServingSQLAddr(), outgoingTLSConfig,
			)
		},
```

The rest of the fields in `BackendConfig` will be removed in a follow up PR once all the uses are migrated to `BackendDialer`

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57431)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @darinpp, @imjching, and @spaskob)

___
*[pkg/ccl/sqlproxyccl/backend_dialer.go, line 20 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOCM9ko2XFyjOcKuWVX:-MOCM9ko2XFyjOcKuWVY:b8odigo) ([raw file](https://github.com/cockroachdb/cockroach/blob/e829abf904ca57f52fbfc292146671a61c39834d/pkg/ccl/sqlproxyccl/backend_dialer.go#L20)):*
> ```Go
> )
> 
> // BackendDial is an example backend dialer that does a TCP/IP connection
> ```

NIT: Period at end of comments, her and elsewhere.
___
*[pkg/ccl/sqlproxyccl/backend_dialer.go, line 35 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOCLdPn4DGlfg6Hg0sj:-MOCLdPn4DGlfg6Hg0sk:b3ssfww) ([raw file](https://github.com/cockroachdb/cockroach/blob/e829abf904ca57f52fbfc292146671a61c39834d/pkg/ccl/sqlproxyccl/backend_dialer.go#L35)):*
> ```Go
> 		return nil, err
> 	}
> 	err = RelayStartupMsg(conn, msg)
> ```

Why not let the proxy relay the startup message? Would we have any cases where we don't want to relay the message, or where we want to send a different message? I can't think of any.
___
*[pkg/ccl/sqlproxyccl/idle_disconnect_connection_test.go, line 18 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOCMkNI7lKtnazTW6pR:-MOCMkNI7lKtnazTW6pS:ba4n60b) ([raw file](https://github.com/cockroachdb/cockroach/blob/e829abf904ca57f52fbfc292146671a61c39834d/pkg/ccl/sqlproxyccl/idle_disconnect_connection_test.go#L18)):*
> ```Go
> 	"github.com/stretchr/testify/require"
> )
> ```

I think we need a unit test here.
___
*[pkg/ccl/sqlproxyccl/proxy.go, line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MNaK17TpqQofWMqeSUM-r1-38:-MOCMyW78_R2KyJ8LnAW:ba09cm9) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9df4da50f3700d77f0978738446b775df073f64/pkg/ccl/sqlproxyccl/proxy.go#L38)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

thanks, make senses now
</blockquote></details>

Let's get rid of this idle timeout field. Just create the idle disconnect overlay in the backend dialer, like we discussed yesterday.


<!-- Sent from Reviewable.io -->

darinpp reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57431)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @darinpp, @imjching, and @spaskob)

___
*[pkg/ccl/sqlproxyccl/backend_dialer.go, line 20 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOCM9ko2XFyjOcKuWVX:-MOD3eR12X1rV8spnfHv:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/e829abf904ca57f52fbfc292146671a61c39834d/pkg/ccl/sqlproxyccl/backend_dialer.go#L20)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: Period at end of comments, her and elsewhere.
</blockquote></details>

done
___
*[pkg/ccl/sqlproxyccl/backend_dialer.go, line 35 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOCLdPn4DGlfg6Hg0sj:-MOD6slh5CaczLEi86Gx:bta83ok) ([raw file](https://github.com/cockroachdb/cockroach/blob/e829abf904ca57f52fbfc292146671a61c39834d/pkg/ccl/sqlproxyccl/backend_dialer.go#L35)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Why not let the proxy relay the startup message? Would we have any cases where we don't want to relay the message, or where we want to send a different message? I can't think of any.
</blockquote></details>

I pulled it out and into the proxy but I think that isn't better. The only case where I think we wouldn't want to relay the startup message right away is if (or when) we implement the stamping of the pods with tenant id via the SQL connection. In that case the first message (or more) may be about specifying the tenant id and the cert to be used. Having the library (proxy) do hardcoded handshake (the startup message) that can't be customized by the library user is very limiting in my view. 
___
*[pkg/ccl/sqlproxyccl/idle_disconnect_connection_test.go, line 18 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOCMkNI7lKtnazTW6pR:-MOD8KmE5PoNvcmu3wA5:b-8b11j9) ([raw file](https://github.com/cockroachdb/cockroach/blob/e829abf904ca57f52fbfc292146671a61c39834d/pkg/ccl/sqlproxyccl/idle_disconnect_connection_test.go#L18)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I think we need a unit test here.
</blockquote></details>

Added
___
*[pkg/ccl/sqlproxyccl/proxy.go, line 38 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MNaK17TpqQofWMqeSUM-r1-38:-MOD8Mx02UGGkaA641R1:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9df4da50f3700d77f0978738446b775df073f64/pkg/ccl/sqlproxyccl/proxy.go#L38)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Let's get rid of this idle timeout field. Just create the idle disconnect overlay in the backend dialer, like we discussed yesterday.
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-:-MOE6E3z2umIiGrHlYxB:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @darinpp, @imjching, and @spaskob)

___
*[pkg/ccl/sqlproxyccl/backend_dialer.go, line 35 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOCLdPn4DGlfg6Hg0sj:-MOE54JU55U3dKn_RS8S:b3ybv9m) ([raw file](https://github.com/cockroachdb/cockroach/blob/e829abf904ca57f52fbfc292146671a61c39834d/pkg/ccl/sqlproxyccl/backend_dialer.go#L35)):*
<details><summary><i>Previously, darinpp wrote…</i></summary><blockquote>

I pulled it out and into the proxy but I think that isn't better. The only case where I think we wouldn't want to relay the startup message right away is if (or when) we implement the stamping of the pods with tenant id via the SQL connection. In that case the first message (or more) may be about specifying the tenant id and the cert to be used. Having the library (proxy) do hardcoded handshake (the startup message) that can't be customized by the library user is very limiting in my view. 
</blockquote></details>

OK, I'm not too opinionated on this point; I'm OK with either way.


<!-- Sent from Reviewable.io -->

spaskob reviewed (changes requested):
> Reviewed 5 of 11 files at r3, 1 of 7 files at r4, 1 of 4 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57431)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @darinpp, @imjching, and @spaskob)

___
*[pkg/ccl/sqlproxyccl/backend_dialer.go, line 35 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOH1Tjv8BiMiEDjxR1q:-MOH1Tjv8BiMiEDjxR1r:b-akkckv) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0ea62679865a65a63a8c1ec34e205d9815592de/pkg/ccl/sqlproxyccl/backend_dialer.go#L35)):*
> ```Go
> 
> // SSLOverlay upgrades the connection to use SSL.
> func SSLOverlay(conn net.Conn, tlsConfig *tls.Config) (net.Conn, error) {
> ```

nit: We should probably call this PGSSLOverlay since this sends a pgSSLRequest to the backend and requires a pgAcceptSSLRequest.
___
*[pkg/ccl/sqlproxyccl/idle_disconnect_connection.go, line 28 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOH2Dr_DJzrSLVh9QTs:-MOH2Dr_DJzrSLVh9QTt:bhw8yfv) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0ea62679865a65a63a8c1ec34e205d9815592de/pkg/ccl/sqlproxyccl/idle_disconnect_connection.go#L28)):*
> ```Go
> }
> 
> var notSupportedError = errors.Errorf(
> ```

nit: we usually use the convention of putting err first, ie "errNotSupported"
___
*[pkg/ccl/sqlproxyccl/metrics.go, line 60 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOH47biASVzTeOGYyKo:-MOH47biASVzTeOGYyKp:bv6odoj) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0ea62679865a65a63a8c1ec34e205d9815592de/pkg/ccl/sqlproxyccl/metrics.go#L60)):*
> ```Go
> 		Name:        "proxy.err.idle_disconnect",
> 		Help:        "Number of disconnects due to idle timeout",
> 		Measurement: "Disconnects",
> ```

nit: Idle Disconnects since we use the same string below
___
*[pkg/ccl/sqlproxyccl/metrics.go, line 66 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOH4R153Utbit_b7tOv:-MOH4R153Utbit_b7tOw:b-uiri4u) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0ea62679865a65a63a8c1ec34e205d9815592de/pkg/ccl/sqlproxyccl/metrics.go#L66)):*
> ```Go
> 		Name:        "proxy.err.client_disconnect",
> 		Help:        "Number of disconnects initiated by clients",
> 		Measurement: "Disconnects",
> ```

we should call this "Client Disconnects" now that there is a different type of disconnect
___
*[pkg/ccl/sqlproxyccl/proxy.go, line 211 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOH6rVC2V8QyCbR85mI:-MOH6rVC2V8QyCbR85mJ:b-8osnqc) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0ea62679865a65a63a8c1ec34e205d9815592de/pkg/ccl/sqlproxyccl/proxy.go#L211)):*
> ```Go
> 		s.metrics.BackendDownCount.Inc(1)
> 		return NewErrorf(
> 			CodeBackendDown, "relaying StartupMessage to target server: %v", err,
> ```

it's nice to be able to see the outgoing address; you can use `crdbConn.RemoteAddr()`.
___
*[pkg/ccl/sqlproxyccl/proxy.go, line 224 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOH8QLQClVQ4cEA5KQT:-MOH8QLQClVQ4cEA5KQU:b-opwpb2) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0ea62679865a65a63a8c1ec34e205d9815592de/pkg/ccl/sqlproxyccl/proxy.go#L224)):*
> ```Go
> 	ctx, cancel := context.WithCancel(context.Background())
> 	defer cancel()
> 	if s.opts.BackendConfigFromParams != nil {
> ```

This approach seems unnecessarily complex to me - can we keep calling BackendConfigFromParams as before (but gated behind the != nil check) and not really touch the current code?

When we decide to remove it, we can remove all BackendConfig code.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57431)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @darinpp, @imjching, and @spaskob)

___
*[pkg/ccl/sqlproxyccl/backend_dialer.go, line 35 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOH1Tjv8BiMiEDjxR1q:-MOHACtw2qcmk-QEQ2Tw:b9bfcjt) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0ea62679865a65a63a8c1ec34e205d9815592de/pkg/ccl/sqlproxyccl/backend_dialer.go#L35)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

nit: We should probably call this PGSSLOverlay since this sends a pgSSLRequest to the backend and requires a pgAcceptSSLRequest.
</blockquote></details>

I like your other naming suggestions, but think we should keep this as `SSLOverlay`. The naming principle: don't quality names unless there are multiple cases. So if we had 2 different ways to establish SSL (say PG and MySql), then we'd want to qualify the name like you suggest. But since we don't, it should just be SSL, since there's only one way to establish this SSL conn.


<!-- Sent from Reviewable.io -->

spaskob reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57431)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @darinpp, @imjching, and @spaskob)

___
*[pkg/ccl/sqlproxyccl/backend_dialer.go, line 35 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOH1Tjv8BiMiEDjxR1q:-MOHC0UiA2WyROZBZ6Rb:bev2mc8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0ea62679865a65a63a8c1ec34e205d9815592de/pkg/ccl/sqlproxyccl/backend_dialer.go#L35)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I like your other naming suggestions, but think we should keep this as `SSLOverlay`. The naming principle: don't quality names unless there are multiple cases. So if we had 2 different ways to establish SSL (say PG and MySql), then we'd want to qualify the name like you suggest. But since we don't, it should just be SSL, since there's only one way to establish this SSL conn.
</blockquote></details>

Sounds good - we should probably reword the comment though to mention that this function assumes a PG connection.


<!-- Sent from Reviewable.io -->

darinpp reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57431)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @imjching, and @spaskob)

___
*[pkg/ccl/sqlproxyccl/backend_dialer.go, line 35 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOCLdPn4DGlfg6Hg0sj:-MOI539qDdUmbmYOX5DL:b-kqvd5p) ([raw file](https://github.com/cockroachdb/cockroach/blob/e829abf904ca57f52fbfc292146671a61c39834d/pkg/ccl/sqlproxyccl/backend_dialer.go#L35)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

OK, I'm not too opinionated on this point; I'm OK with either way.
</blockquote></details>

In that case, I prefer to have it in the dialer as this makes easier the case that I mentioned. 
___
*[pkg/ccl/sqlproxyccl/backend_dialer.go, line 35 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOH1Tjv8BiMiEDjxR1q:-MOI5QPYD4pLurruOUT-:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0ea62679865a65a63a8c1ec34e205d9815592de/pkg/ccl/sqlproxyccl/backend_dialer.go#L35)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

Sounds good - we should probably reword the comment though to mention that this function assumes a PG connection.
</blockquote></details>

done
___
*[pkg/ccl/sqlproxyccl/idle_disconnect_connection.go, line 28 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOH2Dr_DJzrSLVh9QTs:-MOI5SRJ2TqAckaVVJe9:b-n9tkhd) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0ea62679865a65a63a8c1ec34e205d9815592de/pkg/ccl/sqlproxyccl/idle_disconnect_connection.go#L28)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

nit: we usually use the convention of putting err first, ie "errNotSupported"
</blockquote></details>

changed
___
*[pkg/ccl/sqlproxyccl/metrics.go, line 60 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOH47biASVzTeOGYyKo:-MOI5Uyl5LtO5SXeMECp:b-n9tkhd) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0ea62679865a65a63a8c1ec34e205d9815592de/pkg/ccl/sqlproxyccl/metrics.go#L60)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

nit: Idle Disconnects since we use the same string below
</blockquote></details>

changed
___
*[pkg/ccl/sqlproxyccl/metrics.go, line 66 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOH4R153Utbit_b7tOv:-MOI5Vrn2-dQ-u74JefU:b-7xvj6d) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0ea62679865a65a63a8c1ec34e205d9815592de/pkg/ccl/sqlproxyccl/metrics.go#L66)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

we should call this "Client Disconnects" now that there is a different type of disconnect
</blockquote></details>

changed as we
___
*[pkg/ccl/sqlproxyccl/proxy.go, line 63 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MNaK17TpqQofWMqeSUM-r1-63:-MOI5rJkCYXyziVgOLrK:butf06h) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9df4da50f3700d77f0978738446b775df073f64/pkg/ccl/sqlproxyccl/proxy.go#L63)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

I meant we probably need a comment on what Dialer is used when none is provided.
</blockquote></details>

Not providing dialer is temporary until I fix the manged service side. Then there won't be an option to not provide one.
___
*[pkg/ccl/sqlproxyccl/proxy.go, line 211 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOH6rVC2V8QyCbR85mI:-MOI6V_W1eJox92T0Coj:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0ea62679865a65a63a8c1ec34e205d9815592de/pkg/ccl/sqlproxyccl/proxy.go#L211)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

it's nice to be able to see the outgoing address; you can use `crdbConn.RemoteAddr()`.
</blockquote></details>

Fixed.
___
*[pkg/ccl/sqlproxyccl/proxy.go, line 224 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOH8QLQClVQ4cEA5KQT:-MOI6X6JCgTAUrwFZoOT:b-nf0dv2) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0ea62679865a65a63a8c1ec34e205d9815592de/pkg/ccl/sqlproxyccl/proxy.go#L224)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

This approach seems unnecessarily complex to me - can we keep calling BackendConfigFromParams as before (but gated behind the != nil check) and not really touch the current code?

When we decide to remove it, we can remove all BackendConfig code.
</blockquote></details>

I'm not sure what is the complex part of the approach. The only change is that BackendConfigFromParams is now optional so there is check that wasn't here before. And because it is optional - the BackendConfig is optional. Given that we have only two fields that we need - OnConnectionSuccess and KeepAliveLoop - I just put them together under the same if. We can change if you come up with an use case that requires us to not clean these up.


<!-- Sent from Reviewable.io -->

imjching reviewed (commented):
General approach seems fine. My only concern is where we call `ModifyRequestParams` (which should be after dialing, but before writing to the backend).

> Reviewed 2 of 8 files at r2, 2 of 11 files at r3, 2 of 7 files at r4, 3 of 4 files at r5, 6 of 6 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-:-MOI7Obw6bGE-LCflgZT:b-oubo5k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @darinpp, and @spaskob)

___
*[pkg/ccl/sqlproxyccl/backend_dialer.go, line 34 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOI4TPD-ltwPIfx4Hpu:-MOI4TPD-ltwPIfx4Hpv:b-ijbtu9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0ea62679865a65a63a8c1ec34e205d9815592de/pkg/ccl/sqlproxyccl/backend_dialer.go#L34)):*
> ```Go
> }
> 
> // SSLOverlay upgrades the connection to use SSL.
> ```

nit: `... to use SSL if a tls.Config is specified.`
___
*[pkg/ccl/sqlproxyccl/backend_dialer.go, line 36 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOI497v825ovP9NeA7Q:-MOI497w6JklTFl3xiNb:b-aowtri) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0ea62679865a65a63a8c1ec34e205d9815592de/pkg/ccl/sqlproxyccl/backend_dialer.go#L36)):*
> ```Go
> // SSLOverlay upgrades the connection to use SSL.
> func SSLOverlay(conn net.Conn, tlsConfig *tls.Config) (net.Conn, error) {
> 	if tlsConfig != nil {
> ```

nit: We can swap the logic to make it look cleaner, i.e. check the `tlsConfig == nil` case first.
___
*[pkg/ccl/sqlproxyccl/idle_disconnect_connection.go, line 83 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOI4sVOAHHeN7fXTOiD:-MOI4sVOAHHeN7fXTOiE:b-y5400j) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0ea62679865a65a63a8c1ec34e205d9815592de/pkg/ccl/sqlproxyccl/idle_disconnect_connection.go#L83)):*
> ```Go
> 
> // IdleDisconnectOverlay upgrades the connection to one that closes when
> // idle for more than timeout duration.
> ```

nit: a comment stating that if timeout is 0, connection won't be upgraded would be helpful.
___
*[pkg/ccl/sqlproxyccl/proxy.go, line 152 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOI6Ky77Qi0g-DXLhwY:-MOI6Ky77Qi0g-DXLhwZ:b-18eh4o) ([raw file](https://github.com/cockroachdb/cockroach/blob/43bca46230e8e0297dd86814193f3030e768f93c/pkg/ccl/sqlproxyccl/proxy.go#L152)):*
> ```Go
> 	// Add the sniServerName (if used) as parameter
> 	if sniServerName != "" {
> 		msg.Parameters["sni-server"] = sniServerName
> ```

What's the motivation behind this change, or how will we be using this value? 
___
*[pkg/ccl/sqlproxyccl/proxy.go, line 159 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOI6jRq2JVazszIhrW-:-MOI6jRq2JVazszIhrW0:b-rv554i) ([raw file](https://github.com/cockroachdb/cockroach/blob/43bca46230e8e0297dd86814193f3030e768f93c/pkg/ccl/sqlproxyccl/proxy.go#L159)):*
<details><summary><i alt="summary">Quoted 5 lines of code…</i></summary>

> ```
> 	// We should be able to remove this when the all clients switch to
> 	// backend dialer.
> 	if s.opts.ModifyRequestParams != nil {
> 		s.opts.ModifyRequestParams(msg.Parameters)
> 	}
> ```
</details>


This have to come after calling backendDialer. BackendConfigFromParams in the custom suproxy relies on this to remove certain parameters after usage.


<!-- Sent from Reviewable.io -->

darinpp reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57431)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @imjching, and @spaskob)

___
*[pkg/ccl/sqlproxyccl/backend_dialer.go, line 34 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOI4TPD-ltwPIfx4Hpu:-MOIBCiAAf39C58AF2BX:b-7tfmjp) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0ea62679865a65a63a8c1ec34e205d9815592de/pkg/ccl/sqlproxyccl/backend_dialer.go#L34)):*
<details><summary><i>Previously, imjching (Jay Lim) wrote…</i></summary><blockquote>

nit: `... to use SSL if a tls.Config is specified.`
</blockquote></details>

added
___
*[pkg/ccl/sqlproxyccl/backend_dialer.go, line 36 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOI497v825ovP9NeA7Q:-MOIBE1_CM8vbR5Pp4tj:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0ea62679865a65a63a8c1ec34e205d9815592de/pkg/ccl/sqlproxyccl/backend_dialer.go#L36)):*
<details><summary><i>Previously, imjching (Jay Lim) wrote…</i></summary><blockquote>

nit: We can swap the logic to make it look cleaner, i.e. check the `tlsConfig == nil` case first.
</blockquote></details>

done
___
*[pkg/ccl/sqlproxyccl/idle_disconnect_connection.go, line 83 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOI4sVOAHHeN7fXTOiD:-MOIBH8O9mzvdYqJCgQg:b-7tfmjp) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0ea62679865a65a63a8c1ec34e205d9815592de/pkg/ccl/sqlproxyccl/idle_disconnect_connection.go#L83)):*
<details><summary><i>Previously, imjching (Jay Lim) wrote…</i></summary><blockquote>

nit: a comment stating that if timeout is 0, connection won't be upgraded would be helpful.
</blockquote></details>

added
___
*[pkg/ccl/sqlproxyccl/proxy.go, line 152 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOI6Ky77Qi0g-DXLhwY:-MOIBJto5U42IFzbt3d_:bhq8v8p) ([raw file](https://github.com/cockroachdb/cockroach/blob/43bca46230e8e0297dd86814193f3030e768f93c/pkg/ccl/sqlproxyccl/proxy.go#L152)):*
<details><summary><i>Previously, imjching (Jay Lim) wrote…</i></summary><blockquote>

What's the motivation behind this change, or how will we be using this value? 
</blockquote></details>

This is just a way to pass to the dialer, the information that we have sni name. For now that won't do much. When we get to support that - the dialer will look for it and connect to different pods based on that information.
___
*[pkg/ccl/sqlproxyccl/proxy.go, line 159 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/57431#-MOI6jRq2JVazszIhrW-:-MOIBtJY6nxAM4v-z5Z4:b-x7lhge) ([raw file](https://github.com/cockroachdb/cockroach/blob/43bca46230e8e0297dd86814193f3030e768f93c/pkg/ccl/sqlproxyccl/proxy.go#L159)):*
<details><summary><i>Previously, imjching (Jay Lim) wrote…</i></summary><blockquote>

> ```
> 	// We should be able to remove this when the all clients switch to
> 	// backend dialer.
> 	if s.opts.ModifyRequestParams != nil {
> 		s.opts.ModifyRequestParams(msg.Parameters)
> 	}
> ```

This have to come after calling backendDialer. BackendConfigFromParams in the custom suproxy relies on this to remove certain parameters after usage.
</blockquote></details>

moved. Thanks for pointing this out.


<!-- Sent from Reviewable.io -->

darinpp commented:
bors r+
